File: buildin/cv/classification/vgg16_caffe/infer_cpp/src/infer.cpp
#include <gflags/gflags.h>
#include <mm_runtime.h>
#include <cnrt.h>
#include <opencv2/opencv.hpp>

#include <algorithm>
#include <atomic>
#include <chrono>
#include <condition_variable>
#include <fstream>
#include <future>
#include <memory>
#include <string>
#include <utility>
#include <vector>
#include <sys/stat.h>

#include "utils.hpp"
#include "pre_process.hpp"
#include "post_process.hpp"
#include "model_runner.h"

using namespace std;
using namespace cv;

#define MINVALUE(A, B) (A < B ? A : B)

DEFINE_int32(device_id, 0, "The device index of mlu");
DEFINE_string(magicmind_model, "", "The magicmind model path");
DEFINE_int32(batch_size, 8, "batch_size");
DEFINE_string(image_dir, "", "dataset_dir");
DEFINE_string(label_file, "labels.txt", "labels.txt");
DEFINE_string(name_file, "name.txt", "The label path");
DEFINE_string(result_file, "", "The classification results output file");
DEFINE_string(result_label_file, "", "The classification results label output file");
DEFINE_string(result_top1_file, "", "The classification results top1 output file");
DEFINE_string(result_top5_file, "", "The classification results top5 output file");
DEFINE_int32(image_num, 0, "image number");

int main(int argc, char **argv) {
  gflags::ParseCommandLineFlags(&argc, &argv, true);
  TimeCollapse time_vgg16_caffe("infer vgg16_caffe");

  // create an instance of ModelRunner
  auto vgg16_runner = new ModelRunner(FLAGS_device_id, FLAGS_magicmind_model);
  if (!vgg16_runner->Init(FLAGS_batch_size)) {
    SLOG(ERROR) << "Init vgg16 runnner failed.";
    return false;
  }

  // load images
  SLOG(INFO) << "================== Load Images ====================";
  result *test;
  test = LoadImages(FLAGS_image_dir, FLAGS_batch_size, FLAGS_label_file);
  if (test->image_paths.size() == 0) {
    std::cout << "No images found in dir [" << FLAGS_image_dir << "]. Support JPEG jpg.";
    return 0;
  }

  int img_w = 224, img_h = 224, img_c = 3, n_classes = 1000;
  size_t image_num = test->image_paths.size();
  std::map<int, std::string> name_map = load_name(FLAGS_name_file);

  // batch information
  int batch_counter = 0;
  std::vector<std::string> batch_image_name;
  std::vector<cv::Mat> batch_image;
  std::vector<int> batch_image_label;

  Record result_label(FLAGS_result_label_file);
  Record result_top1_file(FLAGS_result_top1_file);
  Record result_top5_file(FLAGS_result_top5_file);
  Record result_file(FLAGS_result_file);

  // allocate host memory for batch preprpcessed data
  auto batch_data = vgg16_runner->GetHostInputData();
  // one batch input data addr offset
  int batch_image_offset = vgg16_runner->GetInputSizes()[0] / FLAGS_batch_size;

  SLOG(INFO) << "Start run...";

  // cambricon-note: if FLAGS_image_num is not set,the image_num is:test->image_paths.size()
  // if set to a positive num, the image_num is the same as FLAGS_image_num
  // if set to a negative num, the image_num is the same as test->image_paths.size()
  if (FLAGS_image_num > 0) {
    image_num = MINVALUE(FLAGS_image_num, image_num);
  }
  size_t rem_image_num = image_num % FLAGS_batch_size;
  SLOG(INFO) << "Total images : " << image_num;

  for (int i = 0; i < image_num; i++) {
    std::string image_name =
        test->image_paths[i].substr(test->image_paths[i].find_last_of('/') + 1, 23);
    int image_label = test->labels[i];
    std::cout << "Inference img : " << image_name << "\t\t\t" << i + 1 << "/" << image_num;
    cv::Mat src_img = cv::imread(test->image_paths[i]);
    // cv::Mat pro_img (img_h, img_w, CV_8UC3);
    // Preprocess(src_img , img_h , img_w , pro_img);
    cv::Mat pro_img = Preprocess(src_img, img_h, img_w);
    batch_image_name.push_back(image_name);
    batch_image_label.push_back(image_label);

    // batching preprocessed data
    memcpy((u_char *)(batch_data[0]) + batch_counter * batch_image_offset, pro_img.data,
           batch_image_offset);

    batch_counter += 1;
    // image_num may not be divisible by FLAGS_batch.
    // real_batch_size records number of images in every loop, real_batch_size may change in the
    // last loop.
    size_t real_batch_size = (i < image_num - rem_image_num) ? FLAGS_batch_size : rem_image_num;
    std::cout << "real_batch_size: " << real_batch_size << std::endl;
    if (batch_counter % real_batch_size == 0) {
      // copy in
      vgg16_runner->H2D();
      // compute
      vgg16_runner->Run(FLAGS_batch_size);
      // copy out
      vgg16_runner->D2H();
      // get model's output addr in host
      auto host_output_ptr = vgg16_runner->GetHostOutputData();

      auto infer_res = (float *)host_output_ptr[0];

      // post process
      for (int j = 0; j < real_batch_size; j++) {
        if (!FLAGS_result_label_file.empty()) {
          result_label.write("[" + std::to_string(i + 1 - real_batch_size + j) +
                                 "]: " + std::to_string(batch_image_label[j]),
                             false);
        }
        std::vector<int> top5 = ArgTopK(infer_res + j * n_classes, n_classes, 5);
        if (!FLAGS_result_top1_file.empty()) {
          result_top1_file.write(
              "[" + std::to_string(i + 1 - real_batch_size + j) + "]: " + std::to_string(top5[0]),
              false);
        }
        result_file.write("top5 result in " + batch_image_name[j] + ":", false);
        for (int k = 0; k < 5; k++) {
          if (!FLAGS_result_top5_file.empty()) {
            result_top5_file.write(
                "[" + std::to_string(i + 1 - real_batch_size + j) + "]: " + std::to_string(top5[k]),
                false);
          }
          if (!FLAGS_result_file.empty()) {
            result_file.write(std::to_string(k) + " [" + name_map[top5[k]] + "]", false);
          }
        }
      }
      batch_counter = 0;
      batch_image_name.clear();
      batch_image_label.clear();
    }
  }

  // destroy resource
  vgg16_runner->Destroy();
  return 0;
}

File: buildin/cv/classification/googlenet_bn_caffe/infer_python/infer.py
import os
from PIL import Image
from tqdm import tqdm
import numpy as np
import magicmind.python.runtime as mm
import argparse
import sys
sys.path.append("..")
from preprocess import preprocess, imagenet_dataset,load_name
sys.path.append("../../../")
from utils import Record

from mm_runner import MMRunner
from logger import Logger

log = Logger()
IMAGE_HEIGHT = 224
IMAGE_WIDTH = 224

parser = argparse.ArgumentParser()
parser.add_argument(
    "--device_id", 
    dest="device_id",  
    type=int, 
    default=0, 
    help="device_id"
)
parser.add_argument(
    "--magicmind_model", 
    dest="magicmind_model", 
    type=str, 
    default="../data/models/googlenet_bn_caffe", 
    help="save mm model to this path"
)
parser.add_argument(
    "--batch_size", 
    dest="batch_size",  
    type=int, 
    default=1, 
    help="batch_size"
)
parser.add_argument(
    "--image_dir", 
    dest="image_dir",  
    type=str, 
    default="/path/to/modelzoo/datasets/imageNet2012/",
    help="imagenet val datasets"
)
parser.add_argument(
    "--image_num", 
    dest="image_num",  
    type=int, 
    default=10, 
    help="image number"
)
parser.add_argument(
    "--name_file", 
    dest="name_file",
    type=str, 
    default="datasets/imagenet/name.txt", 
    help="imagenet name txt"
)
parser.add_argument(
    "--label_file", 
    dest="label_file",
    type=str, 
    default="/path/to/modelzoo/datasets/imageNet2012/labels.txt",
    help="imagenet val label txt"
)
parser.add_argument(
    "--result_file", 
    dest="result_file",
    type=str, 
    default="../data/images/output/infer_result.txt", 
    help="result_file"
)
parser.add_argument(
    "--result_label_file", 
    dest="result_label_file",  
    type=str, 
    default="../data/images/output/eval_labels.txt",
    help="result_label_file"
)
parser.add_argument(
    "--result_top1_file", 
    dest="result_top1_file", 
    type=str, 
    default="../data/images/output/eval_result_1.txt", 
    help="result_top1_file"
)
parser.add_argument(
    "--result_top5_file", 
    dest="result_top5_file",  
    type=str, 
    default="../data/images/output/eval_result_5.txt", 
    help="result_top5_file")

if __name__ == "__main__":
    args = parser.parse_args()
    if not os.path.exists(args.magicmind_model):
        print("please generate magicmind model first!!!")
        exit()
    model = MMRunner(mm_file=args.magicmind_model, device_id=args.device_id)
    name_map = load_name(args.name_file)
    record = Record(args.result_file)
    result_label = Record(args.result_label_file)
    result_top1 = Record(args.result_top1_file)
    result_top5 = Record(args.result_top5_file)

    count = 0
    log.info("Start run ...")
    batch_size = args.batch_size
    image_num = args.image_num

    dataset = imagenet_dataset(
        val_txt = args.label_file, image_file_path = args.image_dir, count = args.image_num
    )
    
    rem_img_num = image_num % batch_size
    img_idx = 0
    batch_counter = 0
    imgs = np.empty([batch_size, IMAGE_HEIGHT, IMAGE_WIDTH, 3])
    labels = np.empty([batch_size])
    
    for img, label in tqdm(dataset, total=args.image_num):
        data = preprocess(img, transpose=False)
        infer_batch_size = (
            batch_size if img_idx < (image_num - rem_img_num) else rem_img_num
        )
        imgs[batch_counter % infer_batch_size, :, :, :] = data
        labels[batch_counter % infer_batch_size] = label
        batch_counter += 1
        img_idx += 1

        if batch_counter % infer_batch_size == 0:
            batch_counter = 0
            inputs = [imgs]

            # inference
            outputs = model(inputs)

            # post-process
            for pred_idx in range(infer_batch_size):
                pred = outputs[0][pred_idx]
                index = pred.argsort()[::-1]

                record.write("top5 result:", False)
                result_label.write("[%d]: %d" % (count, int(labels[pred_idx])), False)
                result_top1.write("[%d]: %d" % (count, index[0]), False)

                for i in range(5):
                    idx = index[i]
                    name = name_map[idx]
                    record.write("%d [%s]" % (i, name), False)
                    result_top5.write("[%d]: %d" % (count, idx), False)
                count += 1
